File: interviews/ForUsAll/Java/SubsetOfCapacity.java
import java.util.*;
public class SubsetOfCapacity{

    public boolean getSubsetOfCapacity(int capacity, int[] weights) {

        ArrayList<ArrayList<Integer>> results = new ArrayList<ArrayList<Integer>>(); 
        ArrayList<Integer> combination = new ArrayList<>();

        boolean res = this.dfs(capacity, weights, 0, 0, combination, results);

        System.out.println(combination);

        return res;
    }

    private boolean dfs(int capacity, 
                    int[]weights, 
                    int index, 
                    int sum,
                    ArrayList<Integer> combination,
                    ArrayList<ArrayList<Integer>> results) {

        
        if (sum % capacity == 0 && sum != 0) {

            results.add(new ArrayList<Integer>(combination));
            return true;

        }

        if (index == weights.length) {
            return false;
        }

        combination.add(weights[index]);
        if (this.dfs(capacity, weights, index + 1, sum + weights[index], combination, results)) 
            return true;
        combination.remove(combination.size() - 1);

        if (this.dfs(capacity, weights, index + 1, sum, combination, results)) 
            return true;

        return false;
    }

    public static void main(String[] args) {

        SubsetOfCapacity soc = new SubsetOfCapacity(); 

        int capacity = 3;
        int[] weights = {1, 3, 5, 21};

        System.out.println(soc.getSubsetOfCapacity(capacity, weights));
    }

}

File: interviews/FUA/Python/subsetOfCapacity.py
class Solution:

    def isSubsetofCapacity(self, capacity, weights):

        weights.sort() 

        combination = []
        running_sum = 0
        index = 0

        return self.dfs(capacity, weights, index, combination, running_sum)

    def dfs(self, capacity, weights, index, combination, running_sum):

        if running_sum % capacity == 0 and running_sum != 0:
            print(combination)
            return True 

        if index == len(weights):
            return False 

        for weight in [0, weights[index]]:

            combination.append(weight)

            if self.dfs(capacity, weights, index + 1, combination, running_sum + weight):
                return True

            combination.pop()

        return False 

        
solution = Solution() 
capacity = 6
weights = [1, 3, 5, 11] 

print(solution.isSubsetofCapacity(capacity, weights))